include/mra/misc/convolutiondata.h

```diff
@@ -0,0 +1,156 @@
+
+  // template <typename T>
+  // struct ConvolutionData {
+  //   Tensor<T, 2> R, T;
+
+  // }
+
+  template <typename T, Dimension NDIM>
+  class Convolution {
+
+    private:
+      size_type K;
+      Level n;
+      int npt;                                      // number of quadrature points
+      T expnt;
+      T coeff;
+      Translation lx;
+      const T* quad_x;                              // quadrature points
+      const T* quad_w;                              // quadrature weights
+      Tensor<T, 3> autocorrcoef;                    // autocorrelation coefficients
+      Tensor<T, 3> c;                               // autocorrelation coefficients
+      std::map<Key<NDIM>, Tensor<T, 2>> rnlijcache; // map for storing rnlij matrices
+      std::map<Key<NDIM>, Tensor<T, 1>> rnlpcache; // map for storing rnlp matrices
+      std::mutex cachemutex;                        // mutex for thread safety
+
+
+      void autoc(){
+        auto c_view = c.current_view();
+        auto autocorr_view = autocorrcoef.current_view();
+        detail::autocorr_get<T>(K, autocorr_view);
+        c_view = 0.0;
+        std::array<Slice,NDIM> slices = {Slice(0, K), Slice(0, K), Slice(0, 2*K)};
+        c_view(slices) = autocorr_view(slices);
+        slices = {Slice(0, K), Slice(0, K), Slice(2*K, 4*K)};
+        c_view(slices) = autocorr_view(slices);
+      }
+
+      // projection of a Gaussian onto double order polynomials
+      Tensor<T, 1> make_rnlp(const Level n, Translation lx) {
+        mra::Key<NDIM> key(n, std::array<Translation, NDIM>({lx}));
+        auto it = rnlpcache.find(key);
+        if (it != rnlpcache.end()) {
+          return std::move(it->second);
+        }
+        Tensor<T, 1> rnlp(2*K);
+        auto rnlp_view = rnlp.current_view();
+        rnlp_view = 0.0;
+
+        if (lx < 0) lx = -lx-1;
+        T scaledcoeff  = coeff*std::pow(0.5, 0.5*n);
+        T beta = expnt * std::pow(T(0.25), T(n));
+        T h = 1.0/std::sqrt(beta);
+        T nbox = 1.0/h;
+        if (nbox < 1) nbox = 1;
+        h = 1.0/nbox;
+
+        T sch = std::abs(scaledcoeff*h);
+
+        T argmax = std::abs(std::log(1e-22/sch));
+
+        for (size_type box=0; box<nbox; ++box){
+          T xlo = box*h + lx;
+          if (beta*xlo*xlo > argmax) break;
+
+          for (size_type i=0; i<npt; ++i){
+            T* phix = new T[2*K];
+            T xx = xlo + h*quad_x[i];
+            T ee = scaledcoeff*std::exp(-beta*xx*xx)*quad_w[i]*h;
+
+            legendre_scaling_functions(xx-lx, 2*K, &phix[0]);
+
+            for (size_type p=0; p<2*K; ++p) {
+              rnlp_view(p) += ee*phix[p];
+            }
+          }
+        }
+        cachemutex.lock();
+        if (rnlpcache.find(key) == rnlpcache.end()) {
+          assert(rnlpcache.find(key) == rnlpcache.end());
+          rnlpcache.emplace(std::move(key), std::move(rnlp));
+        }
+        cachemutex.unlock();
+        it = rnlpcache.find(key);
+        return std::move(it->second);
```

Comment:
The find must be protected by the cachemutex (the internal structure of `std::map` might change if another thread inserts another tensor and then the find breaks)